izpack-core/src/main/java/com/izforge/izpack/core/data/DefaultVariables.java

```diff
@@ -294,89 +296,133 @@ public synchronized void add(DynamicVariable variable)
         dynamicVariables.add(variable);
     }
 
-    private String evaluateVariable(DynamicVariable variable)
-    {
-        String newValue;
-        try
-        {
-            newValue = variable.evaluate(replacer);
-        }
-        catch (IzPackException exception)
-        {
-            throw exception;
-        }
-        catch (Exception exception)
-        {
-            throw new IzPackException("Failed to refresh dynamic variable (" + variable.getName() + ")", exception);
-        }
-        return newValue;
-    }
-
     /**
      * Refreshes dynamic variables.
-     * This method relies on a pre-ordered list @{code dynamicVariables} according to referred variables to
-     * resolve them in the right order.
      *
      * @throws InstallerException if variables cannot be refreshed
      */
     @Override
     public synchronized void refresh() throws InstallerException
     {
         logger.fine("Refreshing dynamic variables");
-
+        Set<DynamicVariable> checkedVariables = new HashSet<DynamicVariable>();
+        Set<String> unsetVariables = new HashSet<String>();
         Set<String> setVariables = new HashSet<String>();
+        // for dependent dynamic variables a size of dynamicVariables.size()+1 would be enough
+        // in case of conditions, which depend on dynamic variables also, we need more iterations
+        // to be on the safe side, we take 10*dynamicVariables.size()+1
+        int maxCount = 10*dynamicVariables.size()+1;
+        int count=maxCount;
+        boolean changed = true;
+        while (changed) {
```

Comment:
> The intention of this branch IZPACK-1406_with_conditions instead of the previous IZPACK-1406 was to get a deterministic single loop refresh() by taking the variables used in conditions into account.
Now you've switched back to the old behaviour here. Probably this is an unintented change? Isnt't it?

Yes and no. I've tested it before in this way, but I'll do some tests without the loop.
Anyway, if the elements will be well-ordered, the only disadvantage left with the loop should be some more lines of code and one comparison. But it will be better readable without it, of course.